crates/worker/src/partition/snapshot_producer.rs

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) 2023 -  Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::path::PathBuf;
+use std::sync::Arc;
+use std::time::SystemTime;
+
+use anyhow::{anyhow, bail};
+use parking_lot::Mutex;
+use tokio::sync::oneshot::Sender;
+use tracing::{info, warn};
+
+use restate_core::TaskKind;
+use restate_partition_store::snapshots::{PartitionSnapshotMetadata, SnapshotFormatVersion};
+use restate_partition_store::PartitionStore;
+use restate_types::config::Configuration;
+use restate_types::identifiers::{PartitionId, SnapshotId};
+use restate_types::live::LiveLoad;
+use restate_types::time::MillisSinceEpoch;
+
+pub struct SnapshotProducer {
+    pub partition_id: PartitionId,
+    pub config: Box<dyn LiveLoad<Configuration> + Send>,
+    latest_snapshot: Arc<Mutex<Option<SnapshotId>>>,
+}
+
+impl SnapshotProducer {
+    pub fn new(partition_id: PartitionId, config: Box<dyn LiveLoad<Configuration> + Send>) -> Self {
+        Self {
+            partition_id,
+            config,
+            latest_snapshot: Arc::new(Mutex::new(None)),
+        }
+    }
+
+    /// Creates a snapshot in a background task, optionally sending a completion notification.
+    pub(crate) fn create_snapshot(
+        &mut self,
+        partition_store: &mut PartitionStore,
+        response_tx: Option<Sender<anyhow::Result<SnapshotId>>>,
+    ) -> anyhow::Result<()> {
+        let partition_id = self.partition_id;
+
+        let arc = self.latest_snapshot.clone();
+        let lock = arc
+            .try_lock_arc()
+            .ok_or(anyhow!("Snapshot already in progress"))?;
+
+        let partition_store = partition_store.clone();
+        let base_dir = self
+            .config
+            .live_load()
+            .partition_store
+            .snapshots_base_dir()
+            .to_path_buf();
+        let config = self.config.live_load();
+        let cluster_name = config.common.cluster_name().to_string();
+        let node_name = config.common.node_name().to_string();
+
+        restate_core::task_center().spawn_child(
+            TaskKind::PartitionSnapshotProducer,
+            "create-snapshot",
+            Some(self.partition_id),
+            async move {
+                let result = Self::write_snapshot(
+                    cluster_name,
+                    node_name,
+                    partition_store,
+                    partition_id,
+                    base_dir,
+                )
+                .await;
+
+                let mut lock = lock;
+                if let Ok(metadata) = &result {
+                    lock.replace(metadata.snapshot_id);
+                }
```

Comment:
This is where I'm using https://amanieu.github.io/parking_lot/parking_lot/struct.MutexGuard.html - could model this with an atomic boolean but I think the lock is more understandable and less error-prone. The downside is that it's an optional feature in parking_lot.

Reply:
If you used the `tokio::sync::Mutex` its MutexGuard is already Send, and will also not block the thread while the lock is held, so it can be used by other tokio tasks.

Reply:
As a general advise. I wouldn't use a Mutex that's held across await points. All sorts of surprises happen when we do this.

A problem I see with this approach is that because you have the value of the latest snapshot id in `latest_snapshot`, it might be tempting for someone later to try and read it from the partition processor. In that case, they would block on the mutex for long. The value in Arc<Mutex<>> usually indicate that the mutex is held for very short period and it's just meant to protect concurrent updates, but that's not true here.

I would suggest `Arc<AtomicBool>` in this case, or use TaskCenter's `spawn_unmanaged()` and use the task handle itself to check if previous task has completed. I like the latter because it'd allow us to implement `Drop` on the task handle (see how it's done in `BackgroundAppender`) to cancel the task if the partition processor is dropped. This might be important in the future when we need to shutdown processors based on control plane commands.

One way to model this, is to spawn one task on startup, and let that managed task loop over incoming requests (actor) instead of creating a new task on every `store_snapshot()` call.

Reply:
I can see how the mutex would be dangerous, but going lock-free has its own potential pitfalls too. Still, `AtomicBool` is ultimately the better fit here - I don't need to retain a reference to the latest snapshot for now so I've eliminated that altogether. It was just there to put some data in the mutex since I had it anyway.

I tried creating a separate task/actor with its own control channel to own the snapshot creation but it seemed like complete overkill, so I've gone back to the simplest thing which is spawn a short-lived task on demand. Right now the snapshot creation operations are extremely short anyway. I'd suggest to keep it simple until we need more sophistication here. Let me know what you think of the current approach.